File: modules/ctx-lite/tests/fixtures/signatures/sample.cpp
/**
 * A comprehensive C++ module for signature extraction testing.
 */
#include <string>
#include <vector>
#include <map>
#include <memory>
#include <iostream>
#include <fstream>
#include <stdexcept>

/**
 * Application configuration class.
 */
class Configuration {
private:
    std::string name_;
    std::string version_;
    bool debug_;
    int timeout_;

public:
    Configuration(const std::string& name, const std::string& version)
        : name_(name), version_(version), debug_(false), timeout_(30000) {}

    std::string getName() const { return name_; }
    std::string getVersion() const { return version_; }
    bool isDebug() const { return debug_; }
    void setDebug(bool enabled) { debug_ = enabled; }

    int getTimeout() const { return timeout_; }
    void setTimeout(int timeout) { timeout_ = timeout; }

    bool validate() const {
        return !name_.empty() && !version_.empty();
    }
};

/**
 * Abstract base class for data processors.
 */
class DataProcessor {
public:
    virtual ~DataProcessor() = default;
    virtual std::string process(const std::string& data) = 0;
    virtual bool validate(const std::string& data) const = 0;
    virtual std::string getName() const = 0;
};

/**
 * Base processor with caching functionality.
 */
class BaseProcessor : public DataProcessor {
protected:
    std::shared_ptr<Configuration> config_;
    std::map<std::string, std::string> cache_;

    virtual std::string doProcess(const std::string& data) = 0;

public:
    explicit BaseProcessor(std::shared_ptr<Configuration> config)
        : config_(config) {}

    std::string process(const std::string& data) override {
        auto it = cache_.find(data);
        if (it != cache_.end()) {
            return it->second;
        }
        std::string result = doProcess(data);
        cache_[data] = result;
        return result;
    }

    void clearCache() {
        cache_.clear();
    }

    std::shared_ptr<Configuration> getConfig() const {
        return config_;
    }
};

/**
 * Processor for text data.
 */
class TextProcessor : public BaseProcessor {
protected:
    std::string doProcess(const std::string& data) override {
        std::string result = data;
        for (char& c : result) {
            c = std::toupper(c);
        }
        return result;
    }

public:
    using BaseProcessor::BaseProcessor;

    bool validate(const std::string& data) const override {
        return !data.empty();
    }

    std::string getName() const override {
        return "TextProcessor";
    }
};

/**
 * Processor for JSON data.
 */
class JsonProcessor : public BaseProcessor {
protected:
    std::string doProcess(const std::string& data) override {
        // Simplified JSON processing
        if (validate(data)) {
            return data; // In real implementation, would format JSON
        }
        throw std::runtime_error("Invalid JSON");
    }

public:
    using BaseProcessor::BaseProcessor;

    bool validate(const std::string& data) const override {
        return data.find('{') != std::string::npos &&
               data.find('}') != std::string::npos;
    }

    std::string getName() const override {
        return "JsonProcessor";
    }
};

/**
 * Factory for creating processor instances.
 */
class ProcessorFactory {
private:
    std::map<std::string, std::function<std::unique_ptr<DataProcessor>(
        std::shared_ptr<Configuration>)>> factories_;

public:
    ProcessorFactory() {
        registerProcessor("text", [](std::shared_ptr<Configuration> config) {
            return std::make_unique<TextProcessor>(config);
        });
        registerProcessor("json", [](std::shared_ptr<Configuration> config) {
            return std::make_unique<JsonProcessor>(config);
        });
    }

    void registerProcessor(
        const std::string& name,
        std::function<std::unique_ptr<DataProcessor>(
            std::shared_ptr<Configuration>)> factory) {
        factories_[name] = factory;
    }

    std::unique_ptr<DataProcessor> createProcessor(
        const std::string& type,
        std::shared_ptr<Configuration> config) const {
        auto it = factories_.find(type);
        if (it == factories_.end()) {
            throw std::runtime_error("Unknown processor type: " + type);
        }
        return it->second(config);
    }
};

/**
 * Main application manager.
 */
class ApplicationManager {
private:
    std::shared_ptr<Configuration> config_;
    std::map<std::string, std::unique_ptr<DataProcessor>> processors_;
    std::unique_ptr<ProcessorFactory> factory_;

    void initializeDefaultProcessors() {
        processors_["text"] = factory_->createProcessor("text", config_);
        processors_["json"] = factory_->createProcessor("json", config_);
    }

public:
    explicit ApplicationManager(std::shared_ptr<Configuration> config)
        : config_(config), factory_(std::make_unique<ProcessorFactory>()) {
        initializeDefaultProcessors();
    }

    void registerProcessor(const std::string& name,
        std::unique_ptr<DataProcessor> processor) {
        processors_[name] = std::move(processor);
    }

    std::string processFile(const std::string& filename,
        const std::string& processorName) {
        auto it = processors_.find(processorName);
        if (it == processors_.end()) {
            throw std::runtime_error("Unknown processor: " + processorName);
        }

        try {
            std::ifstream file(filename);
            if (!file.is_open()) {
                throw std::runtime_error("Cannot open file: " + filename);
            }
            std::string data((std::istreambuf_iterator<char>(file)),
                std::istreambuf_iterator<char>());
            return it->second->process(data);
        } catch (const std::exception& e) {
            if (config_->isDebug()) {
                std::cerr << "Error processing " << filename << ": "
                    << e.what() << std::endl;
            }
            throw;
        }
    }

    DataProcessor* getProcessor(const std::string& name) {
        auto it = processors_.find(name);
        return it != processors_.end() ? it->second.get() : nullptr;
    }

    std::shared_ptr<Configuration> getConfig() const {
        return config_;
    }
};

/**
 * Utility functions for string encoding.
 */
namespace StringUtils {
    std::string encodeString(const std::string& text) {
        std::string result;
        for (char c : text) {
            result += std::to_string(static_cast<int>(c));
            result += '-';
        }
        if (!result.empty()) result.pop_back();
        return result;
    }

    std::string decodeString(const std::string& encoded) {
        std::string result;
        std::string current;
        for (char c : encoded) {
            if (c == '-') {
                result += static_cast<char>(std::stoi(current));
                current.clear();
            } else {
                current += c;
            }
        }
        if (!current.empty()) {
            result += static_cast<char>(std::stoi(current));
        }
        return result;
    }
}

/**
 * Main entry point.
 */
int main() {
    auto config = std::make_shared<Configuration>("demo", "1.0");
    config->setDebug(true);

    ApplicationManager manager(config);
    std::cout << "C++ Application initialized successfully" << std::endl;

    return 0;
}

File: modules/ctx-lite/tests/fixtures/signatures/sample.java
/**
 * A comprehensive Java module for signature extraction testing.
 */
package com.example.signatures;

import java.io.*;
import java.util.*;
import java.util.concurrent.*;
import com.fasterxml.jackson.databind.ObjectMapper;

/**
 * Application configuration class.
 */
public class Configuration {
    private String name;
    private String version;
    private boolean debug;
    private int timeout;

    public Configuration(String name, String version) {
        this.name = name;
        this.version = version;
        this.debug = false;
        this.timeout = 30000;
    }

    public String getName() {
        return name;
    }

    public String getVersion() {
        return version;
    }

    public boolean isDebug() {
        return debug;
    }

    public void setDebug(boolean debug) {
        this.debug = debug;
    }

    public int getTimeout() {
        return timeout;
    }

    public void setTimeout(int timeout) {
        this.timeout = timeout;
    }

    public boolean validate() {
        return name != null && !name.isEmpty()
            && version != null && !version.isEmpty();
    }
}

/**
 * Interface for data processors.
 */
public interface DataProcessor {
    String process(String data) throws IOException;
    boolean validate(String data);
    String getName();
}

/**
 * Base processor class for common functionality.
 */
public abstract class BaseProcessor implements DataProcessor {
    protected Configuration config;
    protected Map<String, String> cache;

    public BaseProcessor(Configuration config) {
        this.config = config;
        this.cache = new ConcurrentHashMap<>();
    }

    public String processWithCache(String data) throws IOException {
        if (cache.containsKey(data)) {
            return cache.get(data);
        }
        String result = doProcess(data);
        cache.put(data, result);
        return result;
    }

    protected abstract String doProcess(String data) throws IOException;

    public void clearCache() {
        cache.clear();
    }

    public Configuration getConfig() {
        return config;
    }
}

/**
 * Processor for text files.
 */
public class TextProcessor extends BaseProcessor {
    public TextProcessor(Configuration config) {
        super(config);
    }

    @Override
    public String process(String data) throws IOException {
        return processWithCache(data);
    }

    @Override
    protected String doProcess(String data) {
        return data.toUpperCase();
    }

    @Override
    public boolean validate(String data) {
        return data != null && data.length() > 0;
    }

    @Override
    public String getName() {
        return "TextProcessor";
    }
}

/**
 * Processor for JSON files.
 */
public class JsonProcessor extends BaseProcessor {
    private ObjectMapper mapper;

    public JsonProcessor(Configuration config) {
        super(config);
        this.mapper = new ObjectMapper();
    }

    @Override
    public String process(String data) throws IOException {
        return processWithCache(data);
    }

    @Override
    protected String doProcess(String data) throws IOException {
        Object obj = mapper.readValue(data, Object.class);
        return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(obj);
    }

    @Override
    public boolean validate(String data) {
        try {
            mapper.readValue(data, Object.class);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    @Override
    public String getName() {
        return "JsonProcessor";
    }

    public List<String> extractKeys(String data) throws IOException {
        Map<?, ?> map = mapper.readValue(data, Map.class);
        return new ArrayList<>(map.keySet().stream()
            .map(Object::toString)
            .toList());
    }
}

/**
 * Factory for creating processor instances.
 */
public class ProcessorFactory {
    private static final Map<String, Class<? extends DataProcessor>> PROCESSORS =
        Map.ofEntries(
            Map.entry("text", TextProcessor.class),
            Map.entry("json", JsonProcessor.class)
        );

    public static DataProcessor createProcessor(String type, Configuration config)
        throws InstantiationException, IllegalAccessException {
        Class<? extends DataProcessor> clazz = PROCESSORS.get(type);
        if (clazz == null) {
            throw new IllegalArgumentException("Unknown processor type: " + type);
        }
        return clazz.getDeclaredConstructor(Configuration.class).newInstance(config);
    }
}

/**
 * Main application manager.
 */
public class ApplicationManager {
    private Configuration config;
    private Map<String, DataProcessor> processors;
    private ExecutorService executor;

    public ApplicationManager(Configuration config) {
        this.config = config;
        this.processors = new ConcurrentHashMap<>();
        this.executor = Executors.newFixedThreadPool(4);
        initializeDefaultProcessors();
    }

    private void initializeDefaultProcessors() {
        try {
            processors.put("text", ProcessorFactory.createProcessor("text", config));
            processors.put("json", ProcessorFactory.createProcessor("json", config));
        } catch (Exception e) {
            if (config.isDebug()) {
                e.printStackTrace();
            }
        }
    }

    public void registerProcessor(String name, DataProcessor processor) {
        processors.put(name, processor);
    }

    public String processFile(String filename, String processorName) throws IOException {
        DataProcessor processor = processors.get(processorName);
        if (processor == null) {
            throw new IllegalArgumentException("Unknown processor: " + processorName);
        }
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            StringBuilder data = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                data.append(line).append("\n");
            }
            return processor.process(data.toString());
        } catch (IOException e) {
            if (config.isDebug()) {
                System.err.printf("Error processing %s: %s%n", filename, e.getMessage());
            }
            throw e;
        }
    }

    public DataProcessor getProcessor(String name) {
        return processors.get(name);
    }

    public Configuration getConfig() {
        return config;
    }

    public void shutdown() {
        executor.shutdown();
    }
}

/**
 * Utility class for string operations.
 */
public class StringUtils {
    public static String encodeString(String text) {
        StringBuilder result = new StringBuilder();
        for (char c : text.toCharArray()) {
            result.append(String.format("%x", (int) c));
        }
        return result.toString();
    }

    public static String decodeString(String encoded) throws NumberFormatException {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < encoded.length(); i += 2) {
            String hex = encoded.substring(i, i + 2);
            result.append((char) Integer.parseInt(hex, 16));
        }
        return result.toString();
    }
}

/**
 * Main entry point.
 */
public class Main {
    public static void main(String[] args) {
        Configuration config = new Configuration("demo", "1.0");
        config.setDebug(true);
        ApplicationManager manager = new ApplicationManager(config);
        System.out.println("Java Application initialized");
        manager.shutdown();
    }
}
